lib/resty/core/socket.lua

```diff
@@ -107,17 +138,142 @@ local function setoption(cosocket, option, value)
+        return nil, ffi_str(err, errlen[0])
     end
 
     return true
 end
 
 
+local errmsg             = base.get_errmsg_ptr()
+local session_ptr        = ffi_new("void *[1]")
+local server_name_str    = ffi_new("ngx_str_t[1]")
+local openssl_error_code = ffi_new("int[1]")
+
+
+local function setclientcert(cosocket, cert, pkey)
+    if not cert and not pkey then
+        cosocket[SOCKET_CLIENT_CERT_INDEX] = nil
+        cosocket[SOCKET_CLIENT_PKEY_INDEX] = nil
+        return true
+    end
+
+    if not cert or not pkey then
+        return nil,
+               "client certificate must be supplied with corresponding " ..
+               "private key"
+    end
+
+    if type(cert) ~= "cdata" then
+        return nil, "bad client cert type"
+    end
+
+    if type(pkey) ~= "cdata" then
+        return nil, "bad client pkey type"
+    end
+
+    cosocket[SOCKET_CLIENT_CERT_INDEX] = cert
+    cosocket[SOCKET_CLIENT_PKEY_INDEX] = pkey
+
+    return true
+end
+
+
+local function sslhandshake(cosocket, reused_session, server_name, ssl_verify,
+    send_status_req, ...)
+
+    local n = select("#", ...)
+    if not cosocket or n > 1 then
+        error("ngx.socket sslhandshake: expecting 1 ~ 5 arguments " ..
+              "(including the object), but seen " .. (cosocket and 5 + n or 0))
+    end
+
+    local r = get_request()
+    if not r then
+        error("no request found", 2)
+    end
+
+    session_ptr[0] = type(reused_session) == "cdata" and reused_session or nil
+
+    if server_name then
+        server_name_str[0].data = server_name
+        server_name_str[0].len = #server_name
+
+    else
+        server_name_str[0].data = nil
+        server_name_str[0].len = 0
+    end
+
+    local u = get_tcp_socket(cosocket)
+
+    local rc = C.ngx_http_lua_ffi_socket_tcp_sslhandshake(r, u,
+                   session_ptr[0],
+                   reused_session ~= false,
+                   server_name_str,
+                   ssl_verify and 1 or 0,
+                   send_status_req and 1 or 0,
+                   cosocket[SOCKET_CLIENT_CERT_INDEX],
+                   cosocket[SOCKET_CLIENT_PKEY_INDEX],
+                   errmsg)
+
+    if rc == FFI_NO_REQ_CTX then
+        error("no request ctx found", 2)
+    end
+
+    if rc == FFI_ERROR then
+        goto failed
+    end
+
+    if rc == FFI_DONE then
+        return reused_session
+    end
+
+    if rc == FFI_OK and reused_session == false then
+        return true
+    end
+
+    if rc == FFI_AGAIN then
+        co_yield()
+    end
+
+    -- (rc == FFI_OK and reused_session ~= false) or
+    -- (rc == FFI_AGAIN and reused_session == cdata/false then co_yield)
+
+    rc = C.ngx_http_lua_ffi_socket_tcp_get_sslhandshake_result(r, u,
+            session_ptr, errmsg, openssl_error_code)
+
+    if rc == FFI_OK then
+        if reused_session == false then
+            return true
+        end
+
+        if session_ptr[0] == nil then
+            return nil
```

Comment:
should return the error at the same time

Reply:
See my reviews, I think we can remove this `if session_ptr[0] == nil then` branch altogether and keep only the `assert` below.

Reply:
From @chronolaw:

@zhuizhuhaomeng @doujiang24 This line is needed, as https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_socket_tcp.c#L1940-L1942 indicates the session could indeed be `NULL` even in case of a successful handshake (e.g. if the server disallows session reuse).

Therefore, the `assert` below is not necessary, as it can not replace the `if` condition here.

Reply:
Actually, I had a closer look at this logic compared to https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_socket_tcp.c#L1935-L1956. The issue is that:

In the original C implementation, when the server did not turn on session reuse, the `sslhandshake` function returned a `lightuserdata` that is `NULL` and did not set the `__gc` meta method on it. This is different than returning a Lua `nil` value which indicates an error.

Therefore, we should mimic this behavior by returning the `session_ptr[0]` `cdata` directly and not call `ffi.gc` on it. From the user's perspective, the handshake is still a success.